### DISCUSSION TITLE
如何使用ch_PP-OCRv4_server_rec高精度模型？

### DISCUSSION THREAD

**Discussion by Xnollis** [PaddlePaddle/PaddleOCR#13768] [Q&A] [answered]
请问一下：如何在命令行程序`paddleocr`中使用`ch_PP-OCRv4_server_rec`高精度模型？
换句话说，就是：需要在命令行程序`paddleocr`中，使用何种程序参数才可以自动下载并使用`ch_PP-OCRv4_server_rec`(和对应的det)高精度模型？
（目前`paddleocr`已经可以在gpu上成功地运行默认模型并进行文字识别。并且是在首次运行后，自动下载了默认的`ch_PP-OCRv4_rec`以及对应的默认det模型）

运行环境：windows10 x64, Python 3.10
paddle包：paddlepaddle_gpu-2.6.1，paddleocr-2.8.1

CUDA检测命令：
python -c "import paddle; print(paddle.version.cuda())"
11.8
python -c "import paddle; print(paddle.version.cudnn())"
8.6.0

PS：这一个问题解决后，我再研究如何使用Python代码来调用模型 ：）

**Comment by Xnollis** [comment]
我尝试自己手工下载了ch_PP-OCRv4_server_rec高精度模型包（以及DET模型包），解压并替换原有的位于“.paddleocr”文件夹中的`ch_PP-OCRv4_rec`文件夹下的`inference.pdmodel`等三个文件之后，

执行命令`paddleocr --image_dir frame_1.png --use_gpu True`会导致程序报错：

`ExternalError: CUDNN error(9), CUDNN_STATUS_NOT_SUPPORTED.
      [Hint: 'CUDNN_STATUS_NOT_SUPPORTED'.  The functionality requested is not presently supported by cuDNN.  ] (at C:\home\workspace\Paddle\paddle\phi\kernels\fusion\gpu\fused_conv2d_add_act_kernel.cu:610)
      [operator < fused_conv2d_add_act > error]`

但如果运行参数改为`--use_gpu False`，则可以正常得到识别结果。

请问这个问题如何解决？如何正常使用ch_PP-OCRv4_server_rec高精度模型？

**Reply by GreatV** [reply]
你遇到的错误是 CUDNN_STATUS_NOT_SUPPORTED，这通常与 CUDA 和 cuDNN 的版本兼容性有关。

**Reply by Xnollis** [reply]
我从官方网址（https://developer.nvidia.com/cuda-toolkit-archive）分别安装了 CUDA11.7和CUDA11.8，
然后从官方网址（https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/windows-x86_64/）下载并安装了最新的cudnn组件cudnn-windows-x86_64-8.9.7.29_cuda11
无论如何排列组合，paddleocr 程序都无法正确执行，得到的出错信息都是一样的。

**Reply by GreatV** [reply]
什么显卡

**Reply by Xnollis** [reply]
Titan RTX 24GB

**Reply by GreatV** [reply]
可以按照这里的教程重新安装一下 https://www.paddlepaddle.org.cn/install/quick?docurl=/documentation/docs/zh/develop/install/pip/windows-pip.html 

然后验证一下是否安装成功
```python
import paddle
paddle.utils.run_check()
```

**Reply by Xnollis** [reply]
好的，非常感谢~~！
目前的运行结果是：
Running verify PaddlePaddle program ...
I0829 20:42:13.811300 10972 program_interpreter.cc:212] New Executor is Running.
W0829 20:42:13.812299 10972 gpu_resources.cc:119] Please NOTE: device: 0, GPU Compute Capability: 7.5, Driver API Version: 12.2, Runtime API Version: 11.8
W0829 20:42:13.995092 10972 gpu_resources.cc:164] device: 0, cuDNN Version: 8.9.
I0829 20:42:17.127631 10972 interpreter_util.cc:624] Standalone Executor is Used.
PaddlePaddle works well on 1 GPU.
PaddlePaddle is installed successfully! Let's start deep learning with PaddlePaddle now.
------------------------------------------
但还是故障依旧。  我等下再去试试安装paddle2.6的使用CUDA12的版本

**Reply by Xnollis** [reply, marked as answer]
已解决！

需要显卡驱动的大版本号 与CUDA运行时库的大版本号保持一致，比如都是12或都是11。 
然后从官网下载最新的cuDNN-8.9.7组件（ https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/windows-x86_64/ ），解压后放到自己程序所在的目录即可。

通过运行下面这个命令，可以看到Driver API Version 和 Runtime API Version。
`python -c "import paddle; paddle.utils.run_check()"`
后得到以下信息：
`Running verify PaddlePaddle program ...
I0829 22:05:53.175015 10208 program_interpreter.cc:212] New Executor is Running.
W0829 22:05:53.176012 10208 gpu_resources.cc:119] Please NOTE: device: 0, GPU Compute Capability: 7.5, "Driver API Version": 12.2, "Runtime API Version": 12.0
W0829 22:05:53.180001 10208 gpu_resources.cc:164] device: 0, cuDNN Version: 8.9.
I0829 22:05:53.862905 10208 interpreter_util.cc:624] Standalone Executor is Used.
PaddlePaddle works well on 1 GPU.
PaddlePaddle is installed successfully! Let's start deep learning with PaddlePaddle now.`

**Comment by GreatV** [comment]
要在 PaddleOCR 命令行中使用 `ch_PP-OCRv4_server_rec` 高精度模型，你可以通过指定模型参数来实现。以下是使用命令行 `paddleocr` 进行高精度模型识别的步骤：

1. 首先，你需要确保已经下载了 `ch_PP-OCRv4_server_rec` 及其对应的检测模型。
2. 运行 PaddleOCR，执行以下命令来运行 OCR 识别：
   ```bash
   paddleocr --use_angle_cls True --lang ch --det_model_dir path/to/ch_PP-OCRv4_server_det --rec_model_dir path/to/ch_PP-OCRv4_server_rec --image_dir path/to/your/image
   ```

   其中，`path/to/your/image` 是你要进行文字识别的图片路径。

**Reply by Xnollis** [reply]
感谢您的回复！但是问题依然存在，并且错误信息没有发生变化。
我的步骤是：
1.先恢复之前被覆盖的模型文件为原样
2.使用命令行：
`paddleocr --use_angle_cls True --lang ch --det_model_dir E:\ocr\det --rec_model_dir E:\ocr\rec --image_dir e:\ocr\img --use_gpu True`

程序运行后，得到的日志和出错信息为：
` ppocr DEBUG: Namespace(help='==SUPPRESS==',
use_gpu=True,
use_xpu=False,
use_npu=False,
use_mlu=False,
ir_optim=True,
use_tensorrt=False,
min_subgraph_size=15,
precision='fp32',
gpu_mem=500,
gpu_id=0,
image_dir='frame_1.png',
page_num=0,
det_algorithm='DB',
det_model_dir='E:\\ocr\\det',
det_limit_side_len=960,
det_limit_type='max',
det_box_type='quad',
det_db_thresh=0.3,
det_db_box_thresh=0.6,
det_db_unclip_ratio=1.5,
max_batch_size=10,
use_dilation=False,
det_db_score_mode='fast',
det_east_score_thresh=0.8,
det_east_cover_thresh=0.1,
det_east_nms_thresh=0.2,
det_sast_score_thresh=0.5,
det_sast_nms_thresh=0.2,
det_pse_thresh=0,
det_pse_box_thresh=0.85,
det_pse_min_area=16,
det_pse_scale=1,
scales=[8,16,32],
alpha=1.0,
beta=1.0,
fourier_degree=5,
rec_algorithm='SVTR_LCNet',
rec_model_dir='E:\\ocr\\rec',
rec_image_inverse=True,
rec_image_shape='3,48,320',
rec_batch_num=6,
max_text_length=25,
rec_char_dict_path='C:\\Users\\xxxxx\\.venv\\lib\\site-packages\\paddleocr\\ppocr\\utils\\ppocr_keys_v1.txt',
use_space_char=True,
vis_font_path='./doc/fonts/simfang.ttf',
drop_score=0.5,
e2e_algorithm='PGNet',
e2e_model_dir=None,
e2e_limit_side_len=768,
e2e_limit_type='max',
e2e_pgnet_score_thresh=0.5,
e2e_char_dict_path='./ppocr/utils/ic15_dict.txt',
e2e_pgnet_valid_set='totaltext',
e2e_pgnet_mode='fast',
use_angle_cls=True,
cls_model_dir='C:\\Users\\xxxxx/.paddleocr/whl\\cls\\ch_ppocr_mobile_v2.0_cls_infer',
cls_image_shape='3,48,192',
label_list=['0','180'],
cls_batch_num=6,
cls_thresh=0.9,
enable_mkldnn=False,
cpu_threads=10,
use_pdserving=False,
warmup=False,
sr_model_dir=None,
sr_image_shape='3,32,128',
sr_batch_num=1,
draw_img_save_dir='./inference_results',
save_crop_res=False,
crop_res_save_dir='./output',
use_mp=False,
total_process_num=1,
process_id=0,
benchmark=False,
save_log_path='./log_output/',
show_log=True,
use_onnx=False,
return_word_box=False,
output='./output',
table_max_len=488,
table_algorithm='TableAttn',
table_model_dir=None,
merge_no_span_structure=True,
table_char_dict_path=None,
layout_model_dir=None,
layout_dict_path=None,
layout_score_threshold=0.5,
layout_nms_threshold=0.5,
kie_algorithm='LayoutXLM',
ser_model_dir=None,
re_model_dir=None,
use_visual_backbone=True,
ser_dict_path='../train_data/XFUND/class_list_xfun.txt',
ocr_order_method=None,
mode='structure',
image_orientation=False,
layout=True,
table=True,
ocr=True,
recovery=False,
use_pdf2docx_api=False,
invert=False,
binarize=False,
alphacolor=(255,255,255),
lang='ch',
det=True,
rec=True,
type='ocr',
savefile=False,
ocr_version='PP-OCRv4',
structure_version='PP-StructureV2')`

`OSError: In user code:

    File "tools/export_model.py", line 288, in <module>
      main()
    File "tools/export_model.py", line 284, in main
      model, arch_config, save_path, logger, input_shape=input_shape)
    File "tools/export_model.py", line 197, in export_single_model
      paddle.jit.save(model, save_path)
    File "/usr/local/lib/python3.7/dist-packages/paddle/fluid/dygraph/jit.py", line 629, in wrapper
      func(layer, path, input_spec, **configs)
    File "/usr/local/lib/python3.7/dist-packages/decorator.py", line 232, in fun
..........
    File "/usr/local/lib/python3.7/dist-packages/paddle/fluid/framework.py", line 3621, in append_op
      attrs=kwargs.get("attrs", None))
    File "/usr/local/lib/python3.7/dist-packages/paddle/fluid/framework.py", line 2635, in __init__
      for frame in traceback.extract_stack():

    ExternalError: CUDNN error(9), CUDNN_STATUS_NOT_SUPPORTED.
      [Hint: 'CUDNN_STATUS_NOT_SUPPORTED'.  The functionality requested is not presently supported by cuDNN.  ] (at C:\home\workspace\Paddle\paddle\phi\kernels\fusion\gpu\fused_conv2d_add_act_kernel.cu:610)
      [operator < fused_conv2d_add_act > error]
`
